### ISSUE TITLE
Feature suggestion: Make Operating System Commands First-Class Citizens?

### ISSUE THREAD

**Issue by catpea** [google/zx#112]
This is a question, it is not meant to upset or overwhelm, or cause trouble. It is only meant to open up the idea to a discussion. It is to reveal to Anton if this is a good idea to consider, some zx versions down the road; or not.

Background:
I have recently [mentioned an observation on creating a Domain Specific Language within zx](https://github.com/google/zx/issues/91#issuecomment-846227995). The idea looks very nice, and perhaps, if there is interest, it should become a side-feature of zx.

But as I considered improving on the chained notation DSL fuhrer, I arrived at the infinitely clever [d3](https://d3js.org/#enter-exit) style notation, more specifically .enter(), .exit(), that would allow for .for(), .if()/.fi(), and lead us even further astray from JavaScript.

By taking the .chained() DSL notation to the maximum, I realized, that what we may need, is something seemingly simpler. The simple presence of JavaScript functions that represent OS commands.

In context of pipes, which is another layer on top of commands (that serves as a good code readability check), we would go from this:

```JavaScript
let {stdout} = await $`echo "hello"`
  .pipe($`awk '{print $1" world"}'`)
  .pipe($`tr '[a-z]' '[A-Z]'`)
assert(stdout === 'HELLO WORLD\n')
```

to something like this (warning untested code, it is here to show the idea):

```JavaScript
let {stdout} = await echo("hello").pipe(awk('{print $1" world"}')).pipe(tr('[a-z]' '[A-Z]'))
```
or

```JavaScript
let {stdout} = await $.pipes(
  echo("hello"),
  awk('{print $1" world"}'),
  tr('[a-z]', '[A-Z]')
); // end pipe
```
I know this is not good news for all, the dotted notation is so pretty by contrast, and the enter()/.exit() spice is just downright fascinating. But let us consider, what the creators of Nodejs would ponder when building on top of the nodejs API. Meaning what is the next layer that makes the API even more convenient (in context of zx).

## Command Factory

Could it be something like this:

```JavaScript

const { echo, awk, tr } = $.factory(...)


```
or maybe

```JavaScript

import { echo, awk, tr } = from 'zx/cowabanga'

```

These functions that are destructured/imported would be pretty much the same function, only the command they run would change.

(note: In the future, maybe we could have some special handling for certain things like awk (maybe), so there is room here for xz built-ins.)

This is not about the creation of thousands of files, just about a simple consistent and uniform wrapper that wraps whatever commands the user requests, of somehow hints on.

To clarify, these commands would not be a method of an object, they would be a top level JavaScript function in the zx scope.

(note: These functions could maybe also come with perks, little extras like .pipe() perhaps, or .env({NODE_ENV: 'debug', DEBUG: '*'}) .nice(), or .redirect('2>&1') or similar.)

Think about the other programmers Larry Wall, Guido van Rossum, Dennis Ritchie, Ken Thompson, [Brian Kernighan](https://www.youtube.com/watch?v=xnCgoEyz31M), and think about Linux users just learning JavaScript and node, what syntax would they expect from zx to get things done quickly:

```JavaScript
let {stdout} = await $`dep deploy --branch=${branch}` 
```

```JavaScript
let {stdout} = await dep(`deploy --branch=${branch}`)
```

TLDR; Should the ```CAT(1)``` command become a JavaScript function cat(), would that make zx better? or meh? And what would be the best way to implement this?

**Comment by antonmedv** [comment]
The scope of the 🐚zx is to add another layer of abstraction on top of the shell to simplify scripting. 

Not to create DSL or another language for scripting. 

But simplify Node.js scripting.

PS There are plenty of DSLs (like https://github.com/shelljs/shelljs) or languages (https://babashka.org/).

**Comment by catpea** [comment]
I hate bothering you, I really want to close this issue, I just want to mention, you can close it for me if you don;t like the following:

I want to let you know that I ran a little test and it turns out that bringing commands in, is just a few lines of code:

```JavaScript
const {echo, cat, tar, foo, zip, anything} = bs.load;

console.log(anything); // [Function (anonymous)]
console.log(bs.load.foo); // [Function (anonymous)]
console.log(echo("bork")); // stdout: bork   <-- Hahaha!
```

This is so tiny, and powerful, I can't bring myself to close this issue, you do it, end it all Anton :) but look:

```JavaScript
const {ls} = bs.load;
console.log( ls('-lh', '/usr') );
```

prints

```shell
stdout: total 392K
dr-xr-xr-x.   2 root root  68K May 23 10:34 bin
drwxr-xr-x.   2 root root 4.0K Jan 26 01:05 games
drwxr-xr-x.  69 root root 4.0K May 15 14:05 include
dr-xr-xr-x.  66 root root  36K May 23 10:34 lib
dr-xr-xr-x. 195 root root 204K May 23 10:34 lib64
drwxr-xr-x.  56 root root  16K May 23 10:34 libexec
drwxr-xr-x.  12 root root 4.0K Jan 26 01:05 local
drwxr-xr-x.   5 root root 4.0K Oct  8  2020 pgsql-13
dr-xr-xr-x.   2 root root  24K May 23 10:34 sbin
drwxr-xr-x. 338 root root  12K May 21 13:26 share
drwxr-xr-x.   4 root root 4.0K Jan 26 01:05 src
lrwxrwxrwx.   1 root root   10 Jan 26 01:05 tmp -> ../var/tmp

```

and the load function is just this:

```JavaScript

import { spawn } from "child_process";

const load = new Proxy({}, {
    get: function get(target, prop) {
      if (typeof prop == 'string') return (...args) => command(prop, args);
    }
});

function command(name, args){ // replace this with zx
  const ls = spawn(name, args);
  ls.stdout.on("data", data => {
      console.log(`stdout: ${data}`);
  });
}

const bs = {load};
export default bs;

```

I already published this in my code, so it is completely out of my system, and I promise not to bother you about extending zx with new stuff. Just close the issue if the code I post there is no good.

Thank You for your patience.

**Comment by antonmedv** [comment]
I think shelljs already doing this :)

**Comment by catpea** [comment]
Shelljs crossed into the belly of the beast and forgot that code is liability: https://github.com/shelljs/shelljs/tree/master/src they made over 30 programs in there.

My snippet above uses the method name to figure out the command to run, it is an infinite factory. The complete program I made over in my world is a cheerful 9 nines of code with the ninth line stating the following:

// You are meant to become a Great Being, stop staring at the source code and cross the Appalachian Trail: https://www.youtube.com/watch?v=hPSvdKTEZug

So, technically, 8 lines.

But I confess, having experienced the functions, I now feel that having JavaScript functions that are actually just decorated spawns, is actually heavier, ```$`a|b` ``` does seem lighter, in deed.

Now that I used them in my program. It is no longer ```a|b``` but now it is ```await bs( cat('package.json'), grep('name') )``` and it is hard to read and confusing at the first glance, in contrast to the CLI.

But even if just to share that shelljs has ~30 modules, and JavaScript proxies can be used to create a one liner "trick" that kind-of-really-does bring in commands into the scope, I don't feel bad for opening this issue anymore.

My way is not cross platform. Shelljs and [cash](https://github.com/dthree/cash/tree/master/src/commands) as well, code a storm for OS cross-compatibility. I mentioned few days ago that Busybox has 300 commands, programming that would take years. I won't question their design decisions.

My way was just about creating that Bash-like feature where you just use a command name, as if it was a built in function.

zx states that "Bash is great, but when it comes to writing scripts, people usually choose a more convenient programming language." and I was just trying to add to that convenience by creating the presence of UNIX commands in zx.

But, you need to go with your vision,
using strings is what made UNIX and the shell awesome.

Plus, you should know that zx is now number 16 on the most starred list of google github things: https://github.com/google?q=&type=&language=&sort=stargazers

You are going to get into top 10 soon!
Way to go! You must celebrate!
